pkg/storage/allocator.go

```diff
@@ -648,11 +648,16 @@ func filterBehindReplicas(
 		// Raft leader. This is rare enough not to matter.
 		return nil
 	}
-	quorumIndex := getQuorumIndex(raftStatus, 0)
+	// NB: We use raftStatus.Commit instead of getQuorumIndex() because the
+	// latter can return a value that is less than the commit index. This is
+	// useful for Raft log truncation which sometimes wishes to keep those
+	// earlier indexes, but not appropriate for determining which nodes are
+	// behind the actual commit index of the range.
 	candidates := make([]roachpb.ReplicaDescriptor, 0, len(replicas))
 	for _, r := range replicas {
 		if progress, ok := raftStatus.Progress[uint64(r.ReplicaID)]; ok {
-			if progress.Match >= quorumIndex {
+			if progress.State == raft.ProgressStateReplicate &&
+				progress.Match >= raftStatus.Commit {
```

Comment:
Since the fields in raft's proto file aren't commented at all, I assume this is the latest index at which a proposal has been confirmed as committed? Rather than the latest index for which the raft leader has proposed something?